include/LinearSolver.h

```diff
@@ -54,6 +55,79 @@ namespace nalu{
 class LinearSolvers;
 class Simulation;
 
+/** LocalGraphArrays is a helper class for building the arrays describing
+ * the local csr graph, rowPointers and colIndices. These arrays are passed
+ * to the TpetraCrsGraph::setAllIndices method. This helper class is used
+ * within nalu's TpetraLinearSystem class.
+ * See unit-tests in UnitTestLocalGraphArrays.C.
+ */
+class LocalGraphArrays {
+public:
+  static const LocalOrdinal INVALID = std::numeric_limits<LocalOrdinal>::max();
+
+  LocalGraphArrays(const Kokkos::View<size_t*,sierra::nalu::DeviceSpace>& rowLengths)
+  : rowPointers(Teuchos::arcp<size_t>(rowLengths.size()+1)),
+    colIndices()
+  {
+    size_t nnz = compute_row_pointers(rowPointers, rowLengths);
+    colIndices = Teuchos::arcp<LocalOrdinal>(nnz);
+    for(int i=0, iend=colIndices.size(); i<iend; ++i) { colIndices[i] = INVALID; }
+  }
+
+  size_t get_row_length(size_t localRow) const { return rowPointers[localRow+1]-rowPointers[localRow]; }
+
+  void insertIndices(size_t localRow, size_t numInds, const LocalOrdinal* inds, int numDof)
+  {
+    LocalOrdinal* row = &colIndices[rowPointers[localRow]];
+    size_t rowLen = rowPointers[localRow+1] - rowPointers[localRow];
+    LocalOrdinal* rowEnd = find_first_invalid(row, rowLen);
+    for(size_t i=0; i<numInds; ++i) {
+      LocalOrdinal* insertPoint = std::lower_bound(row, rowEnd, inds[i]);
+      if (insertPoint <= rowEnd && *insertPoint != inds[i]) {
+        insert(inds[i], numDof, insertPoint, rowEnd+numDof);
+        rowEnd += numDof;
+      }
+    }
+  }
+
+  static size_t compute_row_pointers(Teuchos::ArrayRCP<size_t>& rowPtrs,
```

Comment:
1. Tpetra has functions for computing row offsets.  They are thread-parallel, work on the GPU, etc.  If you don't want to use them, at least use Kokkos::parallel_scan.

2. Don't use ArrayRCP (see above).